cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,329 @@
+    // Only one thread writes output
+    if (0 == threadIdx.x % warpSize) {
+      destination_mask[output_element] = result_mask;
+    }
+
+    destination_row += blockDim.x * gridDim.x;
+    active_threads =
+        __ballot_sync(active_threads, destination_row < num_destination_rows);
+  }
+}
+
+void materialize_bitmask(gdf_column const* left_col,
+                        gdf_column const* right_col,
+                        gdf_column* out_col,
+                        gdf_size_type const* table_indices,
+                        gdf_size_type const* row_indices,
+                        cudaStream_t stream) {
+    constexpr gdf_size_type BLOCK_SIZE{256};
+    cudf::util::cuda::grid_config_1d grid_config {out_col->size, BLOCK_SIZE };
+    // TODO: cant use just one template flag for valids, so need to test for null explicitly
+    // or make the kernel receive just one input col and call it twice one for each input column
+    materialize_merged_bitmask_kernel
+    <<<grid_config.num_blocks, grid_config.num_threads_per_block, 0, stream>>>
+    (reinterpret_cast<bit_mask_t*>(left_col->valid),
+     reinterpret_cast<bit_mask_t*>(right_col->valid),
+     reinterpret_cast<bit_mask_t*>(out_col->valid),
+     out_col->size,
+     table_indices,
+     row_indices);
+
+    CHECK_STREAM(stream);
+}
+
+std::pair<rmm::device_vector<gdf_size_type>, rmm::device_vector<gdf_size_type>>
+generate_merged_indices(device_table const& left_table,
+                        device_table const& right_table,
+                        rmm::device_vector<int8_t> const& asc_desc,
+                        bool nulls_are_smallest,
+                        cudaStream_t stream) {
+
+    const gdf_size_type left_size  = left_table.num_rows();
+    const gdf_size_type right_size = right_table.num_rows();
+    const gdf_size_type total_size = left_size + right_size;
+
+    auto left_side = thrust::make_constant_iterator(static_cast<gdf_size_type>(LEFT_TABLE));
+    auto right_side = thrust::make_constant_iterator(static_cast<gdf_size_type>(RIGHT_TABLE));
+
+    auto left_indices = thrust::make_counting_iterator(static_cast<gdf_size_type>(0));
+    auto right_indices = thrust::make_counting_iterator(static_cast<gdf_size_type>(0));
+
+    auto left_begin_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(left_side, left_indices));
+    auto right_begin_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(right_side, right_indices));
+
+    auto left_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(left_side + left_size, left_indices + left_size));
+    auto right_end_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(right_side + right_size, right_indices + right_size));
+
+    rmm::device_vector<gdf_size_type> out_table_indices(total_size);
+    rmm::device_vector<gdf_size_type> out_row_indices(total_size);
+    auto output_zip_iterator = thrust::make_zip_iterator(thrust::make_tuple(out_table_indices.begin(), out_row_indices.begin()));
```

Comment:
This would be simplified by instead using a vector of `thrust::pair` instead of two independent vectors.

```
using index_type = thrust::pair< side, gdf_size_type > ; // `first` indices left/right side, `second` indicates the row index
thrust::device_vector<index_type> merged_indices(total_size);
```

Reply:
For merge the input should be convertible to output, I tried `thrust::make_zip_iterator(thrust::make_pair(left_side, left_indices))` as input but it doesn't work. Do you know how to initialize the input?

Reply:
I don't think you need a zip iterator anymore, because you're not zipping together two independent iterators. Instead you've got a single iterator iterating across a range of `thrust::pair`.